### ISSUE TITLE
Use of /dev/disk/by-id/ links breaks on devices which don't have such links

### ISSUE THREAD

**Issue by greghensley** [rockstor/rockstor-core#1400]
As of the new feature of using the symlinks in `/dev/disk/by-id/*` for disk identification, it looks like Rockstor's disk scan works by identifying block devices (perhaps via sysfs) and reversing the device name to a by-id link by scanning the list of  reverse each block device.  But, in the case a link doesn't exist, Rockstor shows the `/dev/*` device file in the Disk management screen (and is stored that way in the `name` field of the storageadmin_disk table).

Those devices, without a by-id link, are not usable by Rockstor because the devices are accessed by arbitrarily appending the `name` to "/dev/disk/by-id/", without regard for the source of the name.

Proposed solution: Either the device file must be validated at each access (such as by `stat()`ing the deviced by-id path with fallback to assuming `name` is the absolute path of the /dev/* file) _or_ the source of the `name` (the absolute path of either the by-id link or the original device file) will need to be stored in the storageadmin_disk table and used directly.

**Comment by greghensley** [comment]
I think this issue can be dropped to a low priority, since I witnessed this behavior while using bcache, which is not officially supported by Rockstor.  The issue presents itself because udev (at least in my version of systemd -- 219-19.el7_2.11) only generates by-id links for (1) devices matching one of `sd*`, `sr*`, `cciss*`, `vd*`, `mccblk*`, or `mspblk*`, (2) md and dm devices, and (3) devices which expose a World Wide Name identifier.

I currently work around this limitation with a custom udev rule which captures the bcache device's "cache set" UUID and generates a by-id link from that.  Side note, in the same rule, I also set the ID_SERIAL attribute of the device to that same UUID to get rid of Rockstor's complaint about uniqueness (the bcache driver doesn't export this attribute, for whatever reason).

**Comment by phillxnet** [comment]
@greghensley You bring up some good points (as usual) but I was reluctant to use the full path as this would interfere with url access of the api and complicate what we had in place already; but I did think of it.

The canonical (short temp_name type of sda etc) is an intended fall back when no by-id names are found, and all features that assume by-id path are disabled as a result. If not then we throw away all advantages in using by-id type names anyway. In my dealings with currently supported devices the only time I saw no by-id name available for a device was when there was also no serial available; hence the sense in then restricting the use of such devices in the first place as per the red warning on no serial devices.

Definitely a tricky one and you may find the new wiki entry on Device management in Rockstor interesting in this regard as I go into how this whole mechanism works and to an extend why it was done this way.

https://forum.rockstor.com/t/device-management-in-rockstor/1768

Interestingly enough I did a related uuid substitution for serial on mdadm raid devices when adding this as a capability on system drives. But this can't work unless there is first a serial which to uniquely identify / track devices (as per above document)

Thanks for opening this issue but I'm not sure it's valid given it pertains to unsupported devices and they are by definition not accounted for. However there is more work to be done here and I hope to improve the by-id robustness when I next expect to come across it which should be in issue #550 and so given this issue is now linked to that one (via this reference) the information in it shouldn't be lost. But given it essentially states that devices we don't support arn't managed we are going to be hard pressed to ever resolve it. The move to by-id is essential to manage boot to boot device identification but yes it is early days on this move but all currently supported device mechanisms are currently functional so I don't see this as a viable issue.

However your udev rules would be most useful for others wanting to do the same and would be best shared on the forum in a dedicated thread. We have recently had a very similar issue with the support of nvme devices where appropriate udev rules were required in order to make them appear in a usable fashion. It's one of the disadvantages of the CentOS base but there are stability advantages which it would be nice to keep hold of. If you do post your workaround for bcache do please state the unsupported nature as you have done here. Otherwise we will get many support requests of those who have jury rigged bcache under their btrfs and expect Rockstor to work with it.

The nvme related forum post where the udev rules where developed (towards the end of the thread).
https://forum.rockstor.com/t/unable-to-update-disk-state-for-nvme-device/1567/20

"The issue presents itself because udev (at least in my version of systemd -- 219-19.el7_2.11) only generates by-id links for (1) devices matching one of sd*, sr*, cciss*, vd*, mccblk*, or mspblk*, (2) md and dm devices, and (3) devices which expose a World Wide Name identifier."
Essentially we only currently support devices that our version of udev supports, but as your cunning workaround demonstrates, the manual addition of missing automated rules that newer versions of udev would otherwise generate, are our current workaround. Demonstrated in your workaround and the one presented in the forum thread above on nvme issues. Although there was also a fix needed on our base code for the strange names of nvme, but a pull request is open on that one.

My suggestion is that this issue is fixed upstream in newer versions of udev and so is not a Rockstor issue; ie newer udev would generate appropriate rules for supported devices. Otherwise we don't supported them without a manual rule intervention such as yours.

Well done on the udev rule creation by the way. Do please consider adding to the forum but yes bcache is definitely not considered as supported and I don't myself intend to support it as I would rather await the use of the existing metadata info already present within btrfs metadata to discern relative devices speeds and use that when it eventually arrives.

Thanks for this find and your diligence in reporting. But no serial no go currently and that also means no by-id so the 2 line up. And bcache is not part of the btrfs fs which is developing plenty fast enough on it's own.

So I propose that this issue be closed as 'won't fix' due to the observed behaviour being brought about by unsupported low level intervention at the device level by bcache.

"I think this issue can be dropped to a low priority, since I witnessed this behavior while using bcache, which is not officially supported by Rockstor." Dropping to a low level is not something GitHub caters for unfortunately, means leaving it open and it would then just be clutter assuming my above comments aren't nonsense of course :) meaning it could never be closed, hence my suggestion we address as 'won't fix' and close it. Do you agree?

Oh and thanks for you further contribution, always a pleasure to have people take such a close look at what is done here with a view to improving it. But I don't intend to play a part in Rockstor reverting to temp-names in the db as that was a mistake that is now rectified, and only now occurs when there is no by-id name and hence no other option. Hence no support in that instance, hence big red warning for a reason.

Thanks @greghensley and sorry if this comes across as terse but the wiki entry referenced may be a better place for suggestions on improved design and is intended to encourage that. But also respect that it represents a current state of the code.

**Comment by greghensley** [comment]
@phillxnet, thanks for taking the time for such a thorough response.

_I agree with closing this issue_

My intention wasn't specifically for support of bcache (I don't expect Rockstor to ever support it, officially), but to highlight a weakness in the current by-id implementation.  It looks like #550 mostly already covers it, though.

I do think I misrepresented myself, in some cases, which I'd like to clarify:

> The canonical (short temp_name type of sda etc) is an intended fall back when no by-id names are found, and all features that assume by-id path are disabled as a result.

The use of the canonical name is essentially what I was suggesting, in the absence of a by-id name, but without the feature disablement.  However, at present, those features are "disabled" only in the sense that they are broken, resulting in either half-empty WebUI pages or popups reporting the python traceback.  There are a number of places in the code where the device `name` is blindly appended to `/dev/disk/by-id/'.  Last I checked, this includes when mounting btrfs subvolumes so, while the mount of the bcache-backed btrfs filesystem succeeded, mounting the individual subvols failed with log entries to the effect of "/dev/disk/by-id//dev/bcache0: not found".  I was suggesting replacing these instances of `"/dev/disk/by-id/" + name` I was suggesting replacing these instances with a call to something like:
```python
def get_dev(name):
  try:
    if os.stat("/dev/disk/by-id/" + name):
      return "/dev/disk/by-id/" + name
  except OSError:
    try:
      if os.stat("/dev/" + name)
        return "/dev/" + name
    except OSError:
      return name
```
BUT, I see that get_dev_byid_name already does this, in effect, and reports it via the second index in the returned tuple.

> But I don't intend to play a part in Rockstor reverting to temp-names in the db as that was a mistake that is now rectified, and only now occurs when there is no by-id name and hence no other option.

I think this is in reference to my suggesting of storing the absolute path of the device in the db.  What I meant there was, rather than stat()ing the device file on each access (as described above), a new db field would be added ("abs_path") and `_update_disk_state()` would change from this:
```python
byid_disk_name, is_byid = get_dev_byid_name(d.name, True)
if (Disk.objects.filter(serial=d.serial).exists()):
    dob = Disk.objects.get(serial=d.serial)
    dob.name = byid_disk_name
else:
    dob = Disk(name=byid_disk_name, serial=d.serial)
```
To this (with appropriate changes to the Disk class):
```python
byid_disk_name, is_byid = get_dev_byid_name(d.name, True)
if (Disk.objects.filter(serial=d.serial).exists()):
    dob = Disk.objects.get(serial=d.serial)
    if is_byid:
        dob.name = byid_disk_name
        dob.abs_path = "/dev/disk/by-id/" + byid_disk_name
    else
        dob.name = name
        dob.abs_path = "/dev/" + name
else:
    if is_byid:
        dob = Disk(name=byid_disk_name, abs_path="/dev/disk/by-id/"+byid_disk_name, serial=d.serial)
    else:
        dob = Disk(name=name, abs_path="/dev/"+name, serial=d.serial)
```
After which, `Disk.abs_path` would be used in the code in place of `"/dev/disk/by-id/" + Disk.name`.  Presumably, with the changes planned for #550, something equivalent is already in the works, in order to also support `/dev/mapper/*` device paths.

----

Thanks for the wiki link.  I appreciate the background info.

In the future, should the forums be my first stop for raising concerns, rather than in github issues?

**Comment by phillxnet** [comment]
@greghensley Thanks for the response; I'm afraid I may have replied a we bit more emotionally that I had intended; however your response is excellent.

"However, at present, those features are "disabled" only in the sense that they are broken, resulting in either half-empty WebUI pages or popups reporting the python traceback. "
To this I would suggest that they are broken on un-supported devices but your point still stands ie: that it would be far better for us to elegantly fail over; even on unsupported devices and I concede that. I was actually referring more to the disabling of these devices as part of new pools and in terms of disabling all SMART calls.

As you say the feature to identify failed by-id path retrieval is in there and the intention was to make use of this much as you have suggested and I was aware of the need to pivot these paths when we come to the full disk LUKS support, as you mention; but I was going to do this on the grounds of identify the devices as LUKS, which I believe lsblk can do, and then flagging this and have _update_disk_state() use the new role field to know of the correct path. However, like wise we can also use this same (rather new role field) feature to flag by-id status or otherwise.

As you evidently appreciate these things are rather involved and we chose to move to by-id prior to the LUKS  as on my quick look at LUKS it seemed like it would need more advanced path subsystems than we had. My initial changes are as you appreciate a simple first step move from hard coded /dev/ to hard coded /dev/disk/by-id however we had a greater issue to first resolve ie that of #1043 Which is why I don't want to support the canonical temp_names of 'sdX' type as although they are canonical we need more robust references. And if we attempt to use these temp_names directly then we are not moving forward. I guess that was my main objection. But still as long as we know, via what ever method (say by way of the role field) then we are still in a position to avoid doing the wrong thing.

We have still to receive confirmation of #1043 's fix and the placing of this foundation change was purposefully delayed (for quite some time) in order that it be placed very early in an update channel testing cycle due to the critical nature of the change. But still in order to address #1043 we needed to do this and yes it is a little fledgling but we have to take these changes stead.  And I hope that you appreciate that the source of truth of a dev name is paradoxically the temp_canonical type which leaves us with only the next best thing to be robust to dev name changes: the by-id. Especially given that our use of dev names in a db, which is by it's nature out of date, can't be temporary. So this way we rely on the underlying udev system to keep our by-id names good, that is it's job and ours is to use only by-id names.

But yes we have taken one step of a few that are needed and although the old temp name storage in db didn't invoke the problem behaviour you indicate it was still wrong, hence #1043 . So better to see what going wrong than to simply mix up one device name with another; randomly!! Hence the serial as canonical in the same vein as udev itself.

If your device path changes were to be implemented as is then we would restore the prior (broken) behaviour of dealing with temp names and that is something we have to get away from if we are not to be eternally bugged by race conditions because we can never trust the name we hold in the db for a device. Allowing arbitrary temp names to work again is wrong but as I say we are taking this one step at a time and I whole heatedly agree that we need to accommodate more elegantly for unsupported devices. But likewise we can't re-introduce the problems of storing untrustworthy names in the db and then using them.

This change has taken 3.5 years to arrive at and is going to break stuff somewhere but so far the only thing that has emerged is with bcache and as you see with appropriate udev rules it then worked so I think it's a step in the right direction and I do think we are on the same page. 

As an aside note I think it is questionable design to keep adding more and more fields to a db to address each issue that arises which is why I introduced the role field to cater for such things in a hopefully more elegant fashion. But hopefully you see that if we simply re-enable working with temp_names we also regress to using 'broken' names. But as discussed as long as we are able to know these devices are unreliable named (via an as yet undefined role entry) then that base is covered and so your Disk.abs_path db extension may well be worth the additional complexity of having to keep this up to date as well. Although such a path may also be stored or derived from a role entry for the disk, but this need to be considered some more. And with a focused intent and only in the context of blocking the use of these devices via these unreliable references.  I think initially though that once we have a better role field et al an abs_path field would be a good idea. But we must not re-enable the use of temp names as we are then back where we started prior to #1357 which would be disheartening in the extreme.

I am looking forward to starting the LUKS but as of yet we have 'sub optimal' testing coverage of some very basic btrfs wrappers and really this has to be my next task (baring wiser words from @schakrava or @MFlyer or an emergency patch) so once I've done a stint on building up the unit testing coverage some I will return to this matter and hopefully we can nail this having to deal with essentially miss-configured devices that can't be tracked from one boot to the next, yet that being one of our prime responsibilities. I think it's doable and that the road is ahead but we also have to carefully avoid making the same mistakes again and learning the hard won lessons on the road to date (temp_names I'm looking at you :) ).

"In the future, should the forums be my first stop for raising concerns, rather than in github issues?"
I think that is best judged by yourself and in this instance it initially looked like a bug / feature request but I wrote the wiki entry just the other day to address the pending discussions on this matter so looks like it was just in time. Most obviously issues on GitHub have to be actionable and as focused as possible, hence the 'wont fix' suggestion for an unsupported / mis-configured device suggestion. But the main thing is 2 or more minds are better than one; but only if they communicate so thanks. Given that the forum does have more eyes on it so if you are less sure of an issue then I suppose that might be the better first stop but you have been invaluable in these issues as well: so your call.

Apologies for being a little emotional about all of this, most notably in my first post, but I am quite taken with Rockstor and it's development and it can be a frustrating business when everyone wants everything to be supported and working flawlessly all in one go. Which of course includes myself.

Thanks for the code samples to make your point and taking the time to review the wiki and the code. Don't be surprised to find some of this implemented one way or another in #550, but wise of course to hopefully not repeating the mistakes of times gone by. Hopefully there are more wiki / technical manuals entries to come but everything takes time and sometimes quite a lot of it.

Thanks.
